components/automate-ui/src/app/modules/infra-proxy/infra-environment-constraint/infra-environment-constraint.component.ts

```diff
@@ -0,0 +1,148 @@
+    this.cookbookConstraints.unshift({
+      id: this.cookbookConstraints.length + 1,
+      name: this.name_id,
+      operator: this.operator_id,
+      version: this.constraintFormGroup.controls['version'].value
+    });
+    this.constraintKeys.forEach((element, index) => {
+      if (element === this.name_id) {
+        this.selectedCookbookNames.unshift(element);
+        this.constraintKeys.splice(index, 1);
+      }
+    });
+
+    this.name_id = this.constraintKeys[0];
+    this.operator_id = this.operatorKeys[0];
+
+    this.constraintFormGroup.controls.version.setValue('');
+    this.toUpdateCookbookData(this.cookbookConstraints);
+  }
+
+  public deleteCookbookVersion(cookbookIndex: number, cookbookConstraint: CookbookConstraintGrid) {
+    this.constraintKeys.push(cookbookConstraint.name);
+    this.constraintKeys.sort((a, b) => a < b ? -1 : a > b ? 1 : 0);
+
+    this.selectedCookbookNames.splice(cookbookIndex, 1);
+
+    this.name_id = this.constraintKeys[0];
+    this.cookbookConstraints.splice(cookbookIndex, 1);
+    this.toUpdateCookbookData(this.cookbookConstraints);
+  }
+
+  public handleInput(event: KeyboardEvent): void {
+    if (this.isNavigationKey(event)) {
+      return;
+    }
+    this.conflictError = false;
+  }
+
+  public handleEditName(newName: string, cookbookIndex: number) {
+    let previousName: string;
+
+    previousName = this.toUpdateSelectedCookbookNames(newName, cookbookIndex, previousName);
+
+    this.cookbookConstraints[cookbookIndex].name = newName;
+
+    this.toUpdateCookbookKeys(newName);
+
+    this.constraintKeys.push(previousName);
+    this.constraintKeys.sort((a, b) => a < b ? -1 : a > b ? 1 : 0);
+    this.name_id = this.constraintKeys[0];
+    this.toUpdateCookbookData(this.cookbookConstraints);
+  }
+
+  public handleEditOperator(newCookbookOperator: string, cookbookIndex: number) {
+
+    this.cookbookConstraints[cookbookIndex].operator = newCookbookOperator;
+
+    this.toUpdateCookbookData(this.cookbookConstraints);
+  }
+
+  public handleEditVersion(newCookbookVersion: string, cookbookIndex: number) {
+
+    this.cookbookConstraints[cookbookIndex].version = newCookbookVersion;
+
+    this.toUpdateCookbookData(this.cookbookConstraints);
+  }
+
+  public isSelected(name: string) {
+    return this.selectedCookbookNames.includes(name);
+  }
+
+  private toUpdateCookbookData(cookbookConstraints: CookbookConstraintGrid[]) {
+    this.constraintValues.emit(cookbookConstraints);
+  }
+
+  private toUpdateCookbookKeys(newName: string) {
+    this.constraintKeys.forEach((element, index) => {
+      if (element === newName) {
+        this.constraintKeys.splice(index, 1);
+      }
+    });
+  }
+
+  private toUpdateSelectedCookbookNames(newName: string, cookbookIndex: number,
+    previousName: string) {
+    this.selectedCookbookNames.forEach((element, index) => {
+      if (index === cookbookIndex) {
+        previousName = element;
+        if (newName && newName !== 'undefined') {
```

Comment:
Is this really looking for the string "undefined" ?? Or are you trying to check if newName is undefined? If the latter, that will not work. The non-value `undefined` is falsy, so all you need is this:
```suggestion
        if (newName) {
```
Note that if this was unit-tested, I would not have to ask the question. 😁 
And I am just noticing that this file does not have any unit tests; there is a lot going on here and it definitely needs tests.